src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorSemanticTokensRangeEndpointBenchmark.cs

```diff
@@ -0,0 +1,171 @@
+
+    private DocumentVersionCache VersionCache { get; set; }
+
+    private Uri DocumentUri => DocumentContext.Uri;
+
+    private IDocumentSnapshot DocumentSnapshot => DocumentContext.Snapshot;
+
+    private VersionedDocumentContext DocumentContext { get; set; }
+
+    private Range Range { get; set; }
+
+    private new IRazorLogger Logger { get; set; }
+
+    private ProjectSnapshotManagerDispatcher ProjectSnapshotManagerDispatcher { get; set; }
+
+    private string PagesDirectory { get; set; }
+
+    private string ProjectFilePath { get; set; }
+
+    private string TargetPath { get; set; }
+
+    [Params(0, 100, 1000)]
+    public int NumberOfCsSemanticRangesToReturn { get; set; }
+
+    [GlobalSetup]
+    public async Task InitializeRazorSemanticAsync()
+    {
+        EnsureServicesInitialized();
+
+        var projectRoot = Path.Combine(RepoRoot, "src", "Razor", "test", "testapps", "ComponentApp");
+        ProjectFilePath = Path.Combine(projectRoot, "ComponentApp.csproj");
+        PagesDirectory = Path.Combine(projectRoot, "Components", "Pages");
+        var filePath = Path.Combine(PagesDirectory, $"SemanticTokens.razor");
+        TargetPath = "/Components/Pages/SemanticTokens.razor";
+
+        var documentUri = new Uri(filePath);
+        var documentSnapshot = GetDocumentSnapshot(ProjectFilePath, filePath, TargetPath);
+        var version = 1;
+        DocumentContext = new VersionedDocumentContext(documentUri, documentSnapshot, version);
+        Logger = new NoopLogger();
+        SemanticTokensRangeEndpoint = new SemanticTokensRangeEndpoint();
+
+        var text = await DocumentContext.GetSourceTextAsync(CancellationToken.None).ConfigureAwait(false);
+        Range = new Range
+        {
+            Start = new Position { Line = 0, Character = 0 },
+            End = new Position { Line = text.Lines.Count - 1, Character = text.Lines.Last().Span.Length - 1 }
+        };
+    }
+
+    [Benchmark(Description = "Razor Semantic Tokens Range Endpoint")]
+    public async Task RazorSemanticTokensRangeEndpointRangesAsync()
+    {
+        var textDocumentIdentifier = new TextDocumentIdentifier { Uri = DocumentUri };
+        var cancellationToken = CancellationToken.None;
+        var documentVersion = 1;
+
+        await UpdateDocumentAsync(documentVersion, DocumentSnapshot, cancellationToken).ConfigureAwait(false);
+
+        var request = new SemanticTokensRangeParams { Range = Range, TextDocument = textDocumentIdentifier };
+
+        var languageServer = RazorLanguageServer.GetInnerLanguageServerForTesting();
+        var requestContext = new RazorRequestContext(DocumentContext, Logger, languageServer.GetLspServices());
+
+        ((TestCustomizableRazorSemanticTokensInfoService)RazorSemanticTokenService).RangeNumberToGenerate = NumberOfCsSemanticRangesToReturn;
+
+        await SemanticTokensRangeEndpoint.HandleRequestAsync(request, requestContext, cancellationToken);
```

Comment:
What exactly are you aiming to benchmark here? Seems like there is a lot of code here, so this benchmark is measuring things like switching to our dispatcher thread, adding to our version cache, initializing the language server, and generating fake semantic ranges. Compare that to, for example, our code actions benchmark which is just measuring [how long we take to process code actions](https://github.com/dotnet/razor/blob/main/src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorCodeActionsBenchmark.cs#L142). I would have thought all bar the last line and creation of the request object would move to GlobalSetup.

Is that difference intentional? At a minimum, I think generating the fake data should be moved to GlobalSetup, otherwise the allocation results will be swamped by the creation of thousands of SemanticRange objects

Reply:
I agree with you re: moving much of the code in the benchmark function into GlobalSetup, and generating the fake data. I will update the code.